api/metadata/models.py

```diff
@@ -0,0 +1,126 @@
+from urllib.parse import urlparse
+
+from core.models import AbstractBaseExportableModel
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+
+from organisations.models import Organisation
+from projects.models import Project
+
+from .fields import GenericObjectID
+
+FIELD_VALUE_MAX_LENGTH = 2000
+
+METADATA_SUPPORTED_MODELS = ["environment"]
+
+# A map of model name to a function that takes the object id and returns the organisation_id
+SUPPORTED_REQUIREMENTS_MAPPING = {
+    "environment": {
+        "organisation": lambda org_id: org_id,
+        "project": lambda project_id: Project.objects.get(
+            id=project_id
+        ).organisation.id,
+    }
+}
+
+
+class FieldType(models.TextChoices):
+    INTEGER = "int"
+    STRING = "str"
+    BOOLEAN = "bool"
+    FLOAT = "float"
+    URL = "url"
+
+
+class MetadataField(AbstractBaseExportableModel):
+    """This model represents a metadata field(specific to an organisation) that can be attached to any model"""
+
+    name = models.CharField(max_length=255)
+    type = models.CharField(
+        max_length=255, choices=FieldType.choices, default=FieldType.STRING
+    )
+    description = models.TextField(blank=True, null=True)
+    organisation = models.ForeignKey(Organisation, on_delete=models.CASCADE)
+
+    def is_field_value_valid(self, field_value: str) -> bool:
+        if len(field_value) > FIELD_VALUE_MAX_LENGTH:
+            return False
+        return self.__getattribute__(f"validate_{self.type}")(field_value)
+
+    def validate_int(self, field_value: str):
+        try:
+            int(field_value)
+        except ValueError:
+            return False
+        return True
+
+    def validate_float(self, field_value: str):
+        try:
+            float(field_value)
+        except ValueError:
+            return False
+        return True
+
+    def validate_bool(self, field_value: str):
+        if field_value.lower() in ["true", "false"]:
+            return True
+        return False
+
+    def validate_url(self, field_value: str):
+        try:
+            result = urlparse(field_value)
+            return all([result.scheme, result.netloc])
+        except ValueError:
+            return False
+
+    def validate_str(self, field_value: str):
+        return True
```

Comment:
Do we need to make sure we convert the value to `str` here (and in the other methods?) ?

Reply:
That's not needed since it's a text field and anything can be converted to string, right?